machinery/topology.go

```diff
@@ -111,12 +112,17 @@ func NewTopology(options ...TopologyOptionsFunc) *Topology {
 
 	addPoliciesToGraph(graph, policies)
 
+	var err error
+	if !validate(graph) {
+		err = errors.New("loop detected in graph check linking functions")
+	}
+
 	return &Topology{
 		graph:       graph,
 		objects:     lo.SliceToMap(o.Objects, associateURL[Object]),
 		targetables: lo.SliceToMap(targetables, associateURL[Targetable]),
 		policies:    lo.SliceToMap(policies, associateURL[Policy]),
-	}
+	}, err
```

Comment:
So if a loop is detected, we still return the topology created? I wonder if this is better than returning `nil`. It looks like the error is just informative, not really to protect the functioning of the topology.

Reply:
I believe `*collection[T].Path(from, to Object)` prevents running into infinite loops already by tracking the [visited nodes](https://github.com/Kuadrant/policy-machinery/blob/5fffd8edfbeff6558dc03be268e802ab9ee61671/machinery/topology.go#L302). This is a lazy [alternative](https://en.wikipedia.org/wiki/Topological_sorting#Depth-first_search) to Kahn's algorithm that aims to avoid trouble only in cases where they could happen (i.e. when traversing the graph), instead of preemptively at build time.

If we want to continue allowing building graphs that contain loops (useful for some cases), I guess the preemptive validation should be an option (opt-in or opt-out), and the error be final in case of enforcing a DAG.

Reply:
I think this could also be related to the issue of where I set the panic. The validate function is set after all the nodes have being added to the graph. At that point there is a valid graph but might not be a valid DAG. The reason for returning the graph is if there is an error the user can save the graph and inspect for the errors. 

I am not sure how the `*collection[T].Path(from, to Object)` prevents the creation of loops. Yes, it stops the infinite loops but none of the test cases cover the loops in the graph, so I don't know what would happen. I will look into it.

Reply:
To be clear, `*collection[T].Path(from, to Object)` does not prevent loop in the graph. It prevents infinite loops at runtime when traversing the graph. This is done by keeping track of the visited nodes and [returning](https://github.com/Kuadrant/policy-machinery/blob/5fffd8edfbeff6558dc03be268e802ab9ee61671/machinery/topology.go#L310) before recursion when it visits a node already visited before.

Reply:
Though I guess my main point here was about making the validation optional. IMO, it should be opt-in or opt-out – not sure which one is best – but definitely an option.